core/src/com/unciv/logic/CityDistance.kt

```diff
@@ -0,0 +1,141 @@
+class CityDistanceData {
+
+    @Transient
+    lateinit var game: GameInfo
+
+    companion object {
+        const val IDENTIFIER_ALL_CIVS = "ALL_CIVS"
+        const val IDENTIFIER_MAJOR_CIVS = "MAJOR_CIVS"
+    }
+
+    private var shouldUpdate: Boolean = true
+
+    /** Identifier -> Map (Tile position -> Distance)
+     *  Identifier is either: Civ name, ALL_CIVS or MAJOR_CIVS */
+    private var data: HashMap<String, HashMap<Vector2, CityDistance?>> = HashMap()
+
+    private fun reset() {
+        data = HashMap()
+        data[IDENTIFIER_ALL_CIVS]= HashMap()
+        data[IDENTIFIER_MAJOR_CIVS] = HashMap()
+    }
+
+    private fun resetPlayer(identifier: String) {
+        data[identifier] = HashMap()
+    }
+
+    private fun updateDistanceIfLower(identifier: String, position: Vector2, city: CityInfo, distance: Int) {
+        val currentDistance = data[identifier]!![position]
+        val newDistance = CityDistance(city, distance)
+        data[identifier]!![position] = CityDistance.compare(currentDistance, newDistance)
+    }
+
+    private fun updateDistances(thisTile: TileInfo, city: CityInfo, owner: CivilizationInfo, isMajor: Boolean) {
+
+        val cityTile = city.getCenterTile()
+        val distance = thisTile.aerialDistanceTo(cityTile)
+
+        updateDistanceIfLower(IDENTIFIER_ALL_CIVS, thisTile.position, city, distance)
+        if (isMajor) {
+            updateDistanceIfLower(IDENTIFIER_MAJOR_CIVS, thisTile.position, city, distance)
+            updateDistanceIfLower(owner.civName, thisTile.position, city, distance)
+        }
+    }
+
+    private fun update() {
+
+        // Clear previous info
+        reset()
+
+        for (player in game.civilizations.asSequence()) {
+
+            // Not interested in defeated players
+            if (player.isDefeated())
+                continue
+
+            val isMajor = player.isMajorCiv()
+            if (isMajor)
+                resetPlayer(player.civName)
+
+            // Update distances for every city and every tile
+            for (city in player.cities.asSequence())
+                for (otherTile in game.tileMap.values.asSequence())
+                    updateDistances(otherTile, city, player, isMajor)
+
+        }
+
+        shouldUpdate = false
+
+    }
+
+    fun getClosestCityDistance(tile: TileInfo, player: CivilizationInfo? = null, majorsOnly: Boolean = false) : Int {
+
+        if (shouldUpdate)
+            update()
+
+        return if (player != null && player.isMajorCiv())
+            data[player.civName]!![tile.position]?.distance ?: Int.MAX_VALUE
+        else if (majorsOnly)
+            data[IDENTIFIER_MAJOR_CIVS]!![tile.position]?.distance ?: Int.MAX_VALUE
+        else
+            data[IDENTIFIER_ALL_CIVS]!![tile.position]?.distance ?: Int.MAX_VALUE
+
+    }
+
+    fun getClosestCity(tile: TileInfo, player: CivilizationInfo? = null, majorsOnly: Boolean = false) : CityInfo? {
```

Comment:
Same as above function, we can have just one that returns the citydistancedata directly